esr_key_value_database_intrinsics.md

```diff
@@ -0,0 +1,286 @@
+
+| Return Value | Description |
+|--------------|-------------|
+| -1           | `itr_a`'s key is less than `itr_b`'s key |
+| 0            | `itr_a`'s key is the same as `itr_b`'s key |
+| 1            | `itr_a`'s key is greater than `itr_b`'s key |
+
+If an iterator has status `iterator_end`, then it compares greater than iterators with status
+`iterator_ok`. Two iterators with status `iterator_end` compare equal (0 return value).
+
+### kv_it_key_compare
+
+```c++
+int      kv_it_key_compare(uint32_t itr, const char* key, uint32_t size);
+```
+
+Compare the position of an iterator to a provided key. It aborts the
+transaction if `itr` wasn't returned by `kv_it_create`, was destroyed,
+or has status `iterator_erased`.
+
+| Return Value | Description |
+|--------------|-------------|
+| -1           | `itr`'s key is less than `key` |
+| 0            | `itr`'s key is the same as `key` |
+| 1            | `itr`'s key is greater than `key` |
+
+If `itr` has status `iterator_end`, then it compares greater than `key`, no matter
+what value `key` has.
+
+### kv_it_move_to_end
+
+```c++
+it_stat  kv_it_move_to_end(uint32_t itr);
+```
+
+Move `itr` to out-of-bounds and return the new status (`iterator_end`). It aborts the
+transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+### kv_it_next
+
+```c++
+it_stat  kv_it_next(uint32_t itr);
+```
+
+Move iterator to next position and return its new status. It aborts the
+transaction if `itr` wasn't returned by `kv_it_create`, was destroyed,
+or has status `iterator_erased`.
+
+If `itr`'s status is `iterator_ok`, then this finds the
+next non-deleted key in range. If found, the new status is `iterator_ok`. If
+not found, the new status is `iterator_end`.
+
+If `itr`'s status is `iterator_end` then this finds the first non-deleted key
+in range. If found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_end`.
+
+`kv_it_next` never returns `iterator_erased`.
+
+### kv_it_prev
+
+```c++
+it_stat  kv_it_prev(uint32_t itr);
+```
+
+Move iterator to previous position and return its new status. It aborts the
+transaction if `itr` wasn't returned by `kv_it_create`, was destroyed,
+or has status `iterator_erased`.
+
+If `itr`'s status is `iterator_ok`, then this finds the previous non-deleted
+key in range. If found, the new status is `iterator_ok`. If not found, the
+new status is `iterator_end`.
+
+If `itr`'s status is `iterator_end` then this finds the last non-deleted key
+in range. If found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_end`.
+
+`kv_it_prev` never returns `iterator_erased`.
+
+### kv_it_lower_bound
+
+```c++
+it_stat  kv_it_lower_bound(uint32_t itr, const char* key, uint32_t size);
+```
+
+Find the first non-deleted key >= the provided key and return the new iterator status.
+It aborts the transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
```

Comment:
I think the lookup should also be restricted to the iterator's prefix.  Rationale: no iterator operation should depend on the state of the table outside its prefix.

Reply:
It is. That's one of the restrictions in "An iterator, during its lifetime: ..."

Reply:
At least according to the tests, lower_bound for a key before the prefix can either return end or the first element in range depending on whether the table has other elements between the given key and the prefix.
https://github.com/EOSIO/eos/blob/8b5a9d263a4fa2b2dad6ac6d8609b6ac69dbb6c6/unittests/kv_tests.cpp#L207

I'm arguing that this case should always give the first element in range.

Reply:
Good point. changed.

Reply:
The way you've phrased it is correct, but makes it sound more complex than it is.  I would just say that it's the least key which is >= the provided key and has the correct prefix.

Reply:
changed